File: src/main/java/ludwig/interpreter/Callable.java
package ludwig.interpreter;

public interface Callable {
    default Object call(Object... args) {
        Object result = tail(args);

        if (result instanceof Delayed) {
            return ((Delayed) result).get();
        }

        return result;
    }

    Object tail(Object... args);

    default boolean isLazy() {
        return false;
    }

    int argCount();
}

File: src/main/java/ludwig/interpreter/Callable.kt
package ludwig.interpreter

interface Callable {
    fun call(args: Array<Any?>): Any? {
        val result = tail(args)

        return if (result is Delayed<*>) {
            result.get()
        } else result

    }

    fun tail(args: Array<Any?>): Any?

    val isLazy: Boolean
        get() = false

    fun argCount(): Int
}
